File: project_euler/2.cpp
#include <iostream>
#include <iterator>
#include <map>

using namespace std;

int fibo(int x, map<int, int> &fib)
{
    map<int, int>::iterator retval;
    retval = fib.find(x);
    if (retval != fib.end())
    {
        return retval->second;
    }
    else
    {
        int value = fibo(x - 1, fib) + fibo(x - 2, fib);
        fib.insert(pair<int, int>(x, value));
        return value;
    }
}
int main()
{
    map<int, int> fib;
    fib.insert(pair<int, int>(1, 1));
    fib.insert(pair<int, int>(2, 2));

    int sum = 0;
    int r = 1;

    while (1)
    {
        int t = fibo(r++, fib);
        if (t > 4000000)
            break;
        if (t % 2 == 0)
            sum += t;
    }
    cout << sum << endl;
    return 0;
}

File: project_euler/2.py
def question2():
    fibo = {1: 1, 2: 2}

    def fib(x):
        if x in fibo:
            return fibo[x]
        else:
            fibo[x] = fib(x - 1) + fib(x - 2)
            return fibo[x]

    sum_ = 0
    r = 1
    while True:
        t = fib(r)
        if t > 4000000:
            break
        if t % 2 == 0:
            sum_ += t
        r += 1
    return sum_
